igvm/vm.py

```diff
@@ -208,6 +252,100 @@ def shutdown(self, check_vm_up_on_transaction=True, transaction=None):
                 force_stop_failed=check_vm_up_on_transaction,
             )
 
+    def aws_shutdown(self, timeout: int = 120):
+        """AWS shutdown
+
+        Shutdown a VM in AWS.
+
+        :param: timeout: Timeout value for VM shutdown
+        """
+        ec2 = boto3.client('ec2')
+
+        try:
+            ec2.stop_instances(
+                InstanceIds=[self.dataset_obj['aws_instance_id']],
+                DryRun=True
+            )
+        except ClientError as e:
+            if 'DryRunOperation' not in str(e):
+                raise
+
+        try:
+            response = ec2.stop_instances(
+                InstanceIds=[self.dataset_obj['aws_instance_id']],
+                DryRun=False
+            )
+            current_state = response['StoppingInstances'][0][
+                'CurrentState']['Code']
+            log.info(response)
+            if current_state and current_state == AWS_RETURN_CODES['stopped']:
+                log.info('{} is already stopped.'.format(
+                    self.dataset_obj['hostname']))
+        except ClientError as e:
+            raise VMError(e)
+
+        vm_status_code = self.aws_describe_instance_status(
+            self.dataset_obj['aws_instance_id'])
+        if vm_status_code != AWS_RETURN_CODES['stopped']:
+            log.info('Shutting down "{}" in AWS'.format(
+                self.dataset_obj['hostname']))
+            for retry in range(timeout):
+                log.info('Waiting for VM "{}" to shutdown'.format(
+                    self.dataset_obj['hostname']))
+                vm_status_code = self.aws_describe_instance_status(
+                    self.dataset_obj['aws_instance_id'])
+                if vm_status_code == AWS_RETURN_CODES['stopped']:
+                    log.info('"{}" is stopped.'.format(
+                        self.dataset_obj['hostname']))
+                    break
+                time.sleep(1)
+
+    def aws_describe_instance_status(self, instance_id: str) -> int:
+        """AWS describe instance status
+
+        Get the actual VM status in AWS, e.g. running, stopped or terminated.
+
+        :param: instance_id: Instance ID to get the status for
+
+        :return: return code of instance state as int
+        """
+
+        ec2 = boto3.client('ec2')
+        response = ec2.describe_instances(
+            Filters=[
+                {
+                    'Name': 'instance-state-code',
+                    'Values': [
+                        str(AWS_RETURN_CODES['pending']),
```

Comment:
I know at the end it gets consumed by the AWS web API but isn't it still better to use int here instead of str as we are dealing with numbers.

Reply:
Afair the API call will fail, because, for whatever reason, it's expecting strings as instance-state-code. But the documentation is also telling me integer should be possible. I will double check

Reply:
Yep, not working, thats why we need to use `str` here:
```
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter Filters[0].Values[0], value: 0, type: <class 'int'>, valid types: <class 'str'>
Invalid type for parameter Filters[0].Values[1], value: 16, type: <class 'int'>, valid types: <class 'str'>
Invalid type for parameter Filters[0].Values[2], value: 32, type: <class 'int'>, valid types: <class 'str'>
Invalid type for parameter Filters[0].Values[3], value: 48, type: <class 'int'>, valid types: <class 'str'>
Invalid type for parameter Filters[0].Values[4], value: 64, type: <class 'int'>, valid types: <class 'str'>
Invalid type for parameter Filters[0].Values[5], value: 80, type: <class 'int'>, valid types: <class 'str'>
```